crates/rules/src/roles/abstract_role_builder_or_built.rs

```diff
@@ -0,0 +1,112 @@
+use std::marker::PhantomData;
+
+use serde::{Deserialize, Serialize};
+
+use crate::prelude::*;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct AbstractRoleBuilderOrBuilt<F, T> {
+    #[serde(skip)]
+    #[doc(hidden)]
+    built: PhantomData<T>,
+    role: RoleKind,
+    threshold: u8,
+    threshold_factors: Vec<F>,
+    override_factors: Vec<F>,
+}
+
+pub(crate) type AbstractBuiltRoleWithFactor<F> = AbstractRoleBuilderOrBuilt<F, ()>;
+pub(crate) type RoleBuilder = AbstractRoleBuilderOrBuilt<FactorSourceID, RoleWithFactorSourceIds>;
+
+impl<F, T> AbstractRoleBuilderOrBuilt<F, T> {
+    pub(crate) fn with_factors(
+        role: RoleKind,
+        threshold: u8,
+        threshold_factors: impl IntoIterator<Item = F>,
+        override_factors: impl IntoIterator<Item = F>,
+    ) -> Self {
+        Self {
+            built: PhantomData,
+            role,
+            threshold,
+            threshold_factors: threshold_factors.into_iter().collect(),
+            override_factors: override_factors.into_iter().collect(),
+        }
+    }
+}
+
+impl RoleBuilder {
```

Comment:
neither method on the `RoleBuilder` is meant to be `pub` the idea is that hosts always use a single `MatrixBuilder` (ShieldBuilder)` and which expose the needed APIs.